torchdp/test/virtual_step_test.py

```diff
@@ -0,0 +1,311 @@
+#!/usr/bin/env python3
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+import unittest
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.utils.data import DataLoader
+from torchdp import PrivacyEngine
+from torchvision import transforms
+from torchvision.datasets import FakeData
+
+
+class SampleConvNet(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.conv1 = nn.Conv2d(1, 16, 8, 3)
+        self.conv2 = nn.Conv1d(16, 32, 3, 1)
+        self.convf = nn.Conv1d(32, 32, 1, 1)
+        for p in self.convf.parameters():
+            p.requires_grad = False
+        self.fc1 = nn.Linear(23, 17)
+        self.fc2 = nn.Linear(32 * 17, 10)
+
+    def forward(self, x):
+        # x of shape [B, 1, 28, 28]
+        x = F.relu(self.conv1(x))  # -> [B, 16, 10, 10]
+        x = F.max_pool2d(x, 2, 2)  # -> [B, 16, 5, 5]
+        x = x.view(x.shape[0], x.shape[1], x.shape[2] * x.shape[3])  # -> [B, 16, 25]
+        x = F.relu(self.conv2(x))  # -> [B, 32, 23]
+        x = self.convf(x)  # -> [B, 32, 23]
+        x = self.fc1(x)  # -> [B, 32, 17]
+        x = x.view(-1, x.shape[-2] * x.shape[-1])  # -> [B, 32 * 17]
+        x = self.fc2(x)  # -> [B, 10]
+        return x
+
+    def name(self):
+        return "SampleConvNet"
+
+
+class GradientAccumulation_test(unittest.TestCase):
+    def setUp(self):
+        self.DATA_SIZE = 64
+        self.BATCH_SIZE = 16
+        self.LR = 0  # we want to call optimizer.step() without modifying the model
+        self.ALPHAS = [1 + x / 10.0 for x in range(1, 100, 10)]
+        self.criterion = nn.CrossEntropyLoss()
+
+        self.setUp_data()
+        self.setUp_model_and_optimizer()
+
+    def setUp_data(self):
+        self.ds = FakeData(
+            size=self.DATA_SIZE,
+            image_size=(1, 35, 35),
+            num_classes=10,
+            transform=transforms.Compose(
+                [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
+            ),
+        )
+        self.dl = DataLoader(self.ds, batch_size=self.BATCH_SIZE)
+
+    def setUp_model_and_optimizer(self):
+        self.model = SampleConvNet()
+        self.optimizer = torch.optim.SGD(
+            self.model.parameters(), lr=self.LR, momentum=0
+        )
+
+        self.optimizer.zero_grad()
+
+        # accumulate .grad over the entire dataset
+        for x, y in self.dl:
+            logits = self.model(x)
+            loss = self.criterion(logits, y)
+            loss.backward()
+
+        self.effective_batch_grad = torch.cat(
+            [p.grad.reshape(-1) for p in self.model.parameters() if p.requires_grad]
+        ) * (self.BATCH_SIZE / self.DATA_SIZE)
+
+        self.optimizer.zero_grad()
+
+        # accumulate grads for each pair of batches
+        self.double_grads = []
```

Comment:
Could you please add a comment in places that you use `double_grads`, in that what is the "pair of batches" and why we have a pair?

Reply:
Sorry, these were a relic for a test of the behavior of zero_grad(), which we ultimately changed. I've refactored the test to check that accumulated gradients are correctly erased upon a call to `.step()`. These "double_grads" are no longer needed.